File: yaml_editor.cpp
// yaml_editor.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <cstring>
#include <yaml-cpp/yaml.h>
#include <valijson/adapters/yaml_cpp_adapter.hpp>
#include <valijson/schema.hpp>
#include <valijson/schema_parser.hpp>
#include <valijson/validator.hpp>

#ifdef _WIN32
#include <io.h>
#define isatty _isatty
#define fileno _fileno
#else
#include <unistd.h>
#endif

bool colorEnabled = false;

std::string colorize(const std::string& text, const std::string& color) {
    if (!colorEnabled) return text;
    std::string code;
    if (color == "green") code = "\033[92m";
    else if (color == "red") code = "\033[91m";
    else if (color == "yellow") code = "\033[93m";
    else if (color == "blue") code = "\033[94m";
    else return text;
    return code + text + "\033[0m";
}

std::string readFile(const std::string& path) {
    std::ifstream file(path);
    if (!file) throw std::runtime_error("Cannot open file");
    std::stringstream ss;
    ss << file.rdbuf();
    return ss.str();
}

YAML::Node loadYAML(const std::string& content) {
    try {
        return YAML::Load(content);
    } catch (const YAML::Exception& e) {
        throw std::runtime_error(std::string("YAML syntax error: ") + e.what());
    }
}

std::string formatYAML(const YAML::Node& node) {
    YAML::Emitter emitter;
    emitter << YAML::BeginSeq; // для pretty print можно использовать параметры
    // Но проще вывести как строку через YAML::Dump
    YAML::Node copy = YAML::Clone(node);
    std::stringstream ss;
    ss << copy;
    return ss.str();
}

int main(int argc, char* argv[]) {
    std::string schemaFile;
    bool format = false;
    bool forceColor = false;
    std::string filePath;

    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        if (arg == "-s" || arg == "--schema") {
            if (i + 1 < argc) schemaFile = argv[++i];
            else { std::cerr << "Missing schema file\n"; return 1; }
        } else if (arg == "-F" || arg == "--format") {
            format = true;
        } else if (arg == "-c" || arg == "--color") {
            forceColor = true;
        } else if (arg == "-h" || arg == "--help") {
            std::cout << "Usage: " << argv[0] << " [options] [file]\n";
            std::cout << "  -s, --schema <file>   JSON Schema file\n";
            std::cout << "  -F, --format          pretty-print YAML\n";
            std::cout << "  -c, --color           force color output\n";
            return 0;
        } else {
            if (filePath.empty()) filePath = arg;
            else { std::cerr << "Extra argument: " << arg << "\n"; return 1; }
        }
    }

    colorEnabled = forceColor || isatty(fileno(stdout));

    std::string content;
    if (!filePath.empty()) {
        try { content = readFile(filePath); }
        catch (const std::exception& e) {
            std::cerr << colorize("Error reading file: " + std::string(e.what()), "red") << "\n";
            return 1;
        }
    } else {
        if (isatty(fileno(stdin))) {
            std::cerr << colorize("No input provided. Pipe YAML or pass file.", "red") << "\n";
            return 1;
        }
        std::stringstream ss;
        ss << std::cin.rdbuf();
        content = ss.str();
    }

    YAML::Node data;
    try {
        data = loadYAML(content);
    } catch (const std::exception& e) {
        std::cerr << colorize("❌ " + std::string(e.what()), "red") << "\n";
        return 1;
    }

    if (!schemaFile.empty()) {
        std::string schemaContent;
        try { schemaContent = readFile(schemaFile); }
        catch (const std::exception& e) {
            std::cerr << colorize("Failed to load schema: " + std::string(e.what()), "red") << "\n";
            return 1;
        }
        // Парсим схему (JSON)
        try {
            valijson::Schema schema;
            valijson::SchemaParser parser;
            valijson::adapters::YamlCppAdapter yamlAdapter(data);
            // Но для схемы нужен JSON-парсер – используем valijson с адаптером для YAML
            // Упростим: загружаем схему как YAML (хотя это JSON)
            YAML::Node schemaNode = YAML::Load(schemaContent);
            valijson::adapters::YamlCppAdapter schemaAdapter(schemaNode);
            parser.populateSchema(schemaAdapter, schema);
            valijson::Validator validator(schema);
            if (!validator.validate(yamlAdapter)) {
                std::cerr << colorize("❌ Validation failed.", "red") << "\n";
                return 1;
            }
            std::cout << colorize("✅ YAML is valid according to schema.", "green") << "\n";
        } catch (const std::exception& e) {
            std::cerr << colorize("Schema validation error: " + std::string(e.what()), "red") << "\n";
            return 1;
        }
    } else {
        std::cout << colorize("✅ YAML syntax is valid.", "green") << "\n";
    }

    if (format) {
        std::cout << colorize("📄 Formatted output:", "blue") << "\n";
        std::cout << formatYAML(data);
    }

    return 0;
}

File: yaml_editor.rs
// yaml_editor.rs
use std::env;
use std::fs;
use std::io::{self, Read};
use std::process;
use serde_yaml::{Value, from_str};
use serde_json::{from_str as from_json, to_string_pretty};
use jsonschema::JSONSchema;
use colored::*;

fn load_yaml(content: &str) -> Result<Value, String> {
    from_str(content).map_err(|e| format!("YAML syntax error: {}", e))
}

fn load_schema(file: &str) -> Result<serde_json::Value, String> {
    let content = fs::read_to_string(file).map_err(|e| format!("Cannot read schema: {}", e))?;
    from_json(&content).map_err(|e| format!("Invalid JSON schema: {}", e))
}

fn format_yaml(data: &Value) -> Result<String, String> {
    // Сортировка ключей – serde_yaml по умолчанию не сортирует, но можно через сериализацию в BTreeMap
    // Для простоты оставим как есть
    serde_yaml::to_string(data).map_err(|e| e.to_string())
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let mut schema_file = None;
    let mut format = false;
    let mut color = false;
    let mut file_path = None;

    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "-s" | "--schema" => {
                if i + 1 < args.len() {
                    schema_file = Some(args[i+1].clone());
                    i += 2;
                } else {
                    eprintln!("Missing schema file");
                    process::exit(1);
                }
            }
            "-F" | "--format" => { format = true; i += 1; }
            "-c" | "--color" => { color = true; i += 1; }
            "-h" | "--help" => {
                println!("Usage: {} [options] [file]", args[0]);
                println!("  -s, --schema <file>   JSON Schema file");
                println!("  -F, --format          pretty-print YAML");
                println!("  -c, --color           force color output");
                process::exit(0);
            }
            _ => {
                if file_path.is_none() {
                    file_path = Some(args[i].clone());
                    i += 1;
                } else {
                    eprintln!("Extra argument: {}", args[i]);
                    process::exit(1);
                }
            }
        }
    }

    let use_color = color || atty::is(atty::Stream::Stdout);

    let content = if let Some(path) = file_path {
        fs::read_to_string(&path).unwrap_or_else(|e| {
            eprintln!("{}", format!("Error reading file: {}", e).red());
            process::exit(1);
        })
    } else {
        let mut buffer = String::new();
        if io::stdin().read_to_string(&mut buffer).is_err() || buffer.is_empty() {
            eprintln!("{}", "No input provided. Pipe YAML or pass file.".red());
            process::exit(1);
        }
        buffer
    };

    let data = match load_yaml(&content) {
        Ok(d) => d,
        Err(e) => {
            eprintln!("{}", format!("❌ {}", e).red());
            process::exit(1);
        }
    };

    if let Some(schema_path) = schema_file {
        let schema_json = match load_schema(&schema_path) {
            Ok(s) => s,
            Err(e) => {
                eprintln!("{}", format!("❌ {}", e).red());
                process::exit(1);
            }
        };
        let schema = match JSONSchema::compile(&schema_json) {
            Ok(s) => s,
            Err(e) => {
                eprintln!("{}", format!("❌ Invalid schema: {}", e).red());
                process::exit(1);
            }
        };
        // Преобразуем Value в serde_json::Value для валидации
        let json_val = serde_json::to_value(&data).unwrap_or_else(|_| serde_json::Value::Null);
        let result = schema.validate(&json_val);
        if let Err(errors) = result {
            for err in errors {
                eprintln!("{}", format!("❌ {}", err).red());
            }
            process::exit(1);
        }
        println!("{}", "✅ YAML is valid according to schema.".green());
    } else {
        println!("{}", "✅ YAML syntax is valid.".green());
    }

    if format {
        let formatted = format_yaml(&data).unwrap_or_else(|e| {
            eprintln!("{}", format!("Error formatting: {}", e).red());
            process::exit(1);
        });
        println!("{}", "📄 Formatted output:".blue());
        print!("{}", formatted);
    }
}
